### ISSUE TITLE
clang v19+: unexpected behavior under -fno-honor-infinities and -fno-honor-nans

### ISSUE THREAD

**Issue by l0I0I0l** [llvm/llvm-project#191683]
I closed the [issue](https://github.com/llvm/llvm-project/issues/191630) which was reporting **bug** of floating optimizations.

As [originally](https://github.com/llvm/llvm-project/issues/191630#issuecomment-4230663149) posted:

> But I think clang v19+ is going to optimize it too aggressively under `-ffno-inite-math-only`.
>
> [GCC doc](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html)
> > `-ffinite-math-only`
> >    Allow optimizations for floating-point arithmetic that assume that arguments and results are not NaNs or +-Infs. 
> 
> [clang doc](https://clang.llvm.org/docs/UsersManual.html)
> > `-ffinite-math-only`
> >    Allow floating-point optimizations that assume arguments and results are not NaNs or +-Inf. `-ffinite-math-only` implies:
> >        `-fno-honor-infinities`
> >        `-fno-honor-nans`
> 
> With this flag on, compilers are allowed to optimize expressions with the assumption that no NaNs or +-Infs occurs in the computation. But in my opinion, "totally get values optimized away" is not acceptable. Should only allow "simplify the expressions with the no-honoring assumptions".
> 
> It is dangerous if allow compiler to totally **delete** expressions like `__builtin_inf()`, even without any warnings/errors diagnosis. In such a way, the flag `-ffinite-math-only` is too dangerous and so is `-ffast-math`, which will narrowing the usage of these flags.

The behavior of "remove nan/inf expressions" is not restricted on builtin functions. see [godbolt](https://godbolt.org/z/Mdae9qPG8).
```c++
struct Math
{
	static float inf()
	{
		union { unsigned i; float f; } x = {0x7f800000};
		return (x.f); //return __builtin_inf();
	}
};

//clang v19+ will produce just a "ret" instruction with '-O1 -ffinite-math-only'
//this is unexpected: gcc and clang<=18 does not have this behavior
float asm_inf()
{
	return Math::inf();
}
```

**Comment by shafik** [comment]
So you are type punning using a union which is undefined behavior in C++ (this also happens in C if we modify it appropriately), if we just use a straight float we don't see the same behavior: https://godbolt.org/z/n1KaW4969

```cpp
struct Math
{
	static float inf()
	{
		float x = {0x7f800000};
		return x; 
	}
};

float asm_inf()
{
	return Math::inf();
}
```

but the union type punning obfuscates the example a bit, can you show this behavior w/o the union type punning?

**Comment by l0I0I0l** [comment]
@shafik You will get my point if you replace "`return (x.f);`" with "`return __builtin_inf();`" in [godbolt](https://godbolt.org/z/Mdae9qPG8).

I have to emphasis to clarify my purpose:

> With this flag (`-ffinite-math-only`) on, compilers are allowed to optimize expressions with the assumption that no NaNs or +-Infs occurs in the computation. But **in my opinion**, "totally get values optimized away" is not acceptable. **Should only allow** "simplify the expressions with the no-honoring assumptions".
> 
> It is dangerous if allow compiler to totally delete expressions like __builtin_inf(), even without any warnings/errors diagnosis. In such a way, the flag -ffinite-math-only is **too dangerous** and so is -ffast-math, which will **narrowing the usage** of these flags.

**Comment by arsenm** [comment]
There is a warning for using constant infinities and nans with these flags. There is no semantic distinction between a runtime evaluation and something that happens to be evaluatable at compile time.

**Comment by l0I0I0l** [comment]
It is anyway wrong that a compiler translate "`float func(){ /*any legal C++ code*/ }`" to simply a "`ret`" instruction, because it at least have to put a decent value in the return register. Optimizers are allowed to reduce some expression "`a * b`" to "`c`" but should NOT totally delete the expression.

Semantically in C++, it converts "&lt;`expr`&gt;" to `void`, and then return as `float`, which looks unreasonable.
This is clang v19+ specific behavior on `-ffinite-math-only`.

Infinities/Nans are sometime useful to init a variable for computing min/max values. What I think is, compiler can still optimize expressions under the assumption that **unknown** results/args are always finite, but it should be smart to stop the assumption if the expression is **already known** as non-finite at **compile time**.

So why clang v19+ started to delete the expression totally when it know the expressions known as Infinities/Nans at compile time?

**Comment by keinflue** [comment]
According to the LLVM language reference the `nofpclass` attribute given to the function by -ffinite-math-only causes returning any non-finite value from the function to return a poison value instead.

A poison value is allowed to be replaced by any other value, so storing a poison value can be assumed to have no effect because whatever value was present in memory before the store is a valid result and therefore a store for the return value can be optimized away.

This is the mechanism by which this happens. And my (non-authoritative) understanding of `-ffinite-math-only` was always that it will behave as if non-finite values simply didn't exist in the floating point implementation. So this behavior makes sense to me.

Before Clang-19 InstCombine did not fold a
```llvm
ret float /*infinite constant*/
```
to
```llvm
ret float poison
```
even with the `nofpclass(inf)` attribute. I have not checked which commit changed that.

**Comment by l0I0I0l** [comment]
I have no knowledge of LLVM language. Thanks for the explanations and I start to understand something about this optimizations.

The optimization flag `-ffinite-math-only` can be explained either "**strictly** assume(**all/any** values are finite)" or "**smartly** assume(**unknown** values are finite)". I would prefer the latter because the flag is for a purpose in optimization context, not to describe the total source as "all is finite" which is less sense to me.

When compiler knowns a result is non-finite, the program would be ill-formed if further mark the result as *poison*. So `-Werror=nan-infinity-disabled` should be used for safer compilation, but I just think this make life harder.

Does the `poison` serve for more aggressive optimization or more strict assertion/diagnoses? Life would be much easier if explain `-ffinite-math-only` in a more smart way or just revert to the pre-v19 behavior.

**Comment by arsenm** [comment]
The problem is how do you define "smartly"? There needs to be an internally consistent model that also allows optimizations. The current definition is closer to inf/nan are trap values, which you can reason about without context and do something with.

**Comment by AaronBallman** [comment]
I don't see how the language dialect flag `-ffinite-math-only` can be read as "allow non-finite math". It's a language dialect that sets the expectation that inf and nan can *never* be produced. And that's why we diagnose the code when we can: https://godbolt.org/z/K574Gx6Ma (Important note: this is *not* an optimization flag -- it changes the behavior of conforming code, so it's a language dialect flag. Violating the rules of the dialect are UB and so it's reasonable for the optimizer and the constant expression evaluator to both assume the situation can never come up, but the constant expression evaluator has more ability to diagnose the issue.)

**Comment by AaronBallman** [comment]
CC @jcranmer-intel @andykaylor @zahiraam as folks who have also worked in this area in recent history, in case they have different opinions.

**Comment by l0I0I0l** [comment]
I learned the flag `-ffinite-math-only` at [gcc/Optimize-Options](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html), and keep in mind that its mainly for optimization purpose. Now I learned it is a language dialect flag, as a surprise, but still don't see it in [gcc/C-Dialect-Options](https://gcc.gnu.org/onlinedocs/gcc/C-Dialect-Options.html) or [gcc/c++-Dialect-Options](https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html).

The question is: what's the benefit by putting it in language dialect? If a program is worrying about non-finite values, it should also be worrying about any incorrect values/results, and create test cases for assurance. I don't see much sense to validate some specific values {Nans,+-Inf} at compile-time. If such constants are not intended to be, why not add some proper **static_assert()** instead?

_Text below can be ignored if compiler insist that the flag is a language dialect option._

### Idea for allowing non-finite values under `-ffinite-math-only`
- the core of the doc for this flag is "`assume arguments and results are not NaNs or +-Inf`". By "assume", we can only assume on **unknown** things and accept or make use of **known** things (which can be refered as **facts**). If by intention some fact is unexpected, we can **complain** on it instead of **change** it.
- I propose to deprecate `-fno-honor-infinities` and `-fno-honor-nans`, because if infinites/nans are already known at compile time, we can not change the fact. If the fact is unexpected, should explicitly assert on them.
- Then express `-ffinite-math-only` as `-fno-unknown-infinities -fno-unknown-nans`.

**Comment by jcranmer-intel** [comment]
> The optimization flag `-ffinite-math-only` can be explained either "**strictly** assume(**all/any** values are finite)" or "**smartly** assume(**unknown** values are finite)". I would prefer the latter because the flag is for a purpose in optimization context, not to describe the total source as "all is finite" which is less sense to me.

This is the kind of statement that belies a deep misunderstanding of how compilers work. "Smartly" is too poorly specified to mean anything. But you also need to understand that, from the perspective of the optimizer, *most* values are "unknown." In your original example, bitcasting an integer to a floating-point value is unknown from the perspective of the optimizer, at least until constant-folding happens. Just because something looks obvious to the programmer in their source code does not mean it is obvious to the optimizer in the course of optimization.



> I learned the flag `-ffinite-math-only` at [gcc/Optimize-Options](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html), and keep in mind that its mainly for optimization purpose. Now I learned it is a language dialect flag, as a surprise, but still don't see it in [gcc/C-Dialect-Options](https://gcc.gnu.org/onlinedocs/gcc/C-Dialect-Options.html) or [gcc/c++-Dialect-Options](https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html).

LLVM is not GCC, and does not strictly follow gcc's semantics for optimization flags. At best, gcc's option documentation can be a guide for what LLVM's interpretation may mean, but when LLVM or clang itself is documenting it (and `-ffinite-math-only` is in that category), then GCC's documentation isn't particularly relevant.

> The question is: what's the benefit by putting it in language dialect? If a program is worrying about non-finite values, it should also be worrying about any incorrect values/results, and create test cases for assurance. I don't see much sense to validate some specific values {Nans,+-Inf} at compile-time. If such constants are not intended to be, why not add some proper **static_assert()** instead?

The benefit of `ninf`/`nnan` is that it allows you to assert certain algebraic identities that would not hold true if infinities or NaNs were legal values (e.g., `x - x` => `0`). There are also some people who find it useful to provide inlined math libraries that support code paths to handle infinities, where the users who compile with `-ffinite-math-only` get all of those code paths stripped out for them automatically and concomitant performance boosts.

For a user, compiling with `-ffinite-math-only` (note that this is something a user must explicitly opt in to!) is a *promise* that they're not going to use infinities or NaNs, and as a result of that *promise*, the optimizer is free to do optimizations that would change the behavior of code if infinities or NaNs were present (since the user *promised* they weren't). If you are not willing to *promise* not to use infinity or NaN, then do not compile your code with `-ffinite-math-only`.

**Comment by andykaylor** [comment]
I've argued before (https://discourse.llvm.org/t/propogation-of-fpclass-assumptions-vis-a-vis-fast-math-flags/76554) that the `nnan` and `ninf` are too strong to be useful for anything other than strictly assuming the `NaN` and `inf` will never be encountered, which I think is clearly beyond the intention for which the `-ffinite-math-only` option was originally introduced in GCC. The problem, as discussed in the Discourse topic I just linked, is that it's more-or-less impossible to have well-defined semantics for the "weaker" interpretation ("smartly assume..."). Allowing optimizations on that basis ends up being a "trust me, I know what I'm doing" rule, which is fundamentally at odds with the LLVM design philosophy.

This was the entire reason I introduced the separate `-ffp-model=fast` and `-ffp-model=aggressive` command-line options, with the former not setting `nnan` and `ninf`. We are now optimizing so aggressively on `nnan` and `ninf` that it is only useful in the case where you really do want the strict "finite-math only" assumption, though it is very useful for that.

Unfortunately, to have any kind of compatibility with GCC, `-ffast-math` and `-Ofast` have to map to the aggressive mode.

**Comment by jcranmer-intel** [comment]
I should add that I've considered a weaker flag along the lines of `no-overflow` which would essentially make the result poison iff the strict version of the operation would raise FE_OVERFLOW, FE_DIVBYZERO or maybe FE_INVALID flags. It's a semantically defensible definition, but it's not clear to me if it actually usefully enables optimization, especially without a powerful floating-point range analysis.

**Comment by l0I0I0l** [comment]
>This is the kind of statement that belies a deep misunderstanding of how compilers work. "Smartly" is too poorly specified to mean anything. But you also need to understand that, from the perspective of the optimizer, most values are "unknown." In your original example, bitcasting an integer to a floating-point value is unknown from the perspective of the optimizer, at least until constant-folding happens. Just because something looks obvious to the programmer in their source code does not mean it is obvious to the optimizer in the course of optimization.

Well, compiler obviously know something on the moment it mark a result as _poison_. However, it is implementation dependent whether it is easy to control the action of the optimizer.

I visited that impressive [discourse link](https://discourse.llvm.org/t/propogation-of-fpclass-assumptions-vis-a-vis-fast-math-flags/76554), and I really agree with that jyknight said:
>So, personally, I’m definitely in favor of making -ffast-math have less UB in it. I don’t think anyone expects to get UB from enabling that flag, and I don’t think the performance benefits of creating UB are worth the cost.

I would expect optimizer to detect UB instead of introducing it (e.g. mark some value as _poison_ just because optimizer find it violates some condition to perform some optimization pass).

>The problem is how do you define "smartly"? There needs to be an internally consistent model that also allows optimizations. The current definition is closer to inf/nan are trap values, which you can reason about without context and do something with.

I understand it may not be easy to change the current definitions on `-ffinite-math-only`, which may cause a lot change in the implementation. But is it possible to introduce some flags to **control** the UB, so that would keep some balance with the situation that optimizers are being developed in a more aggressive direction. For example, `-f[no-]ub-disable-poison`?
